frontend/catalyst/jax_primitives.py

```diff
@@ -1657,29 +1657,51 @@ def _hamiltonian_lowering(jax_ctx: mlir.LoweringRuleContext, coeffs: ir.Value, *
 def _sample_abstract_eval(obs, shots, shape):
     assert isinstance(obs, AbstractObs)
 
+    """
+    # TODO: restructure these checks now that `shots` is a proper argument instead of an attribute
+    # Is `shape` even necessary now?
     if obs.primitive is compbasis_p:
         assert shape == (shots, obs.num_qubits)
     else:
         assert shape == (shots,)
+    """
 
-    return core.ShapedArray(shape, jax.numpy.float64)
+    # return core.ShapedArray(shape, jax.numpy.float64)
+    return core.DShapedArray(shape, jax.numpy.dtype("float64"))
 
 
 @sample_p.def_impl
 def _sample_def_impl(ctx, obs, shots, shape):  # pragma: no cover
     raise NotImplementedError()
 
 
-def _sample_lowering(jax_ctx: mlir.LoweringRuleContext, obs: ir.Value, shots: int, shape: tuple):
+def _sample_lowering(
+    jax_ctx: mlir.LoweringRuleContext, obs: ir.Value, shots: ir.Value, shape: tuple
+):
     ctx = jax_ctx.module_context.context
     ctx.allow_unregistered_dialects = True
 
     i64_type = ir.IntegerType.get_signless(64, ctx)
-    shots_attr = ir.IntegerAttr.get(i64_type, shots)
+    # breakpoint()
+    shots_value = ConstantOp(i64_type, shots) if isinstance(shots, int) else shots
```

Comment:
There shouldn't be this division. All `shots` use cases should be a proper argument instead of a param.

Reply:
We have adopted a new design: dynamic shots will be an argument, and static shots will remain as a plain i64 literal in the op attribute.